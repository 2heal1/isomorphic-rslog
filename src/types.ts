export type LogLevel = 'log' | 'debug' | 'info' | 'warn' | 'error';

export type LogMessage = number | string | Error | null;

export type LogFormatter = (
  input: string | number | null | undefined,
) => string;

export interface LogType {
  label?: string;
  level: LogLevel;
  LogFormatter?: LogFormatter;
}

export type LogFunction = (message?: LogMessage, ...args: any[]) => void;

export interface Options<T extends Record<string, LogType>> {
  level?: string;
  types?: T;
}
