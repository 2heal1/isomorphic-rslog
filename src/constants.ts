export let LOG_LEVEL = {
  error: 0,
  warn: 1,
  info: 2,
  log: 3,
  verbose: 4,
} as const;
