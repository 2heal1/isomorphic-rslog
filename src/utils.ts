const errorStackRegExp = /^\s*at\s.*:\d+:\d+[\s)]*$/;
const anonymousErrorStackRegExp = /^\s*at\s.*\(<anonymous>\)$/;

export const isErrorStackMessage = (message: string) =>
  errorStackRegExp.test(message) || anonymousErrorStackRegExp.test(message);

export function getLongestLabelLength(labels: (string | undefined)[]) {
  let longestLabel = '';
  labels.forEach(label => {
    if (label && label.length > longestLabel.length) {
      longestLabel = label;
    }
  });
  return longestLabel.length;
}
