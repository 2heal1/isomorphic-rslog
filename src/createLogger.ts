import { LOG_LEVEL, } from './constants';
import { isErrorStackMessage } from './utils';
import type { Options, LogMessage, Logger, LogMethods, LogTypes } from './types';

import type { gradient as TGradient } from './browser/gradient';
import type { finalLog as TFinalLog, getLabel as TGetLabel } from './browser/utils';


export let createLogger = (options: Options = {}, { getLabel, handleError, finalLog, greet, LOG_TYPES }: {
  LOG_TYPES: LogTypes;
  getLabel: typeof TGetLabel;
  finalLog: typeof TFinalLog;
  gradient: typeof TGradient,
  greet: (msg: string) => unknown,
  handleError: (msg: string) => string,

}) => {
  let maxLevel = options.level || 'log';
  const customLabels = options.labels || {};

  let log = (type: LogMethods, message?: LogMessage, ...args: string[]) => {
    if (LOG_LEVEL[LOG_TYPES[type].level] > LOG_LEVEL[maxLevel]) {
      return;
    }

    if (message === undefined || message === null) {
      return console.log();
    }

    let logType = LOG_TYPES[type];
    let text = '';

    const label = getLabel(type, logType, customLabels);

    if (message instanceof Error) {
      if (message.stack) {
        let [name, ...rest] = message.stack.split('\n');
        if (name.startsWith('Error: ')) {
          name = name.slice(7);
        }
        text = `${name}\n${handleError(rest.join('\n'))}`;
      } else {
        text = message.message;
      }
    }

    // change the color of error stacks to
    else if (logType.level === 'error' && typeof message === 'string') {

      let lines = message.split('\n');
      text = lines
        .map(line => (isErrorStackMessage(line) ? handleError(line) : line))
        .join('\n');
    } else {
      text = `${message}`;
    }

    finalLog(label, text, args, message)
  };

  let logger = {
    // greet
    greet: (message: string) => log('log', greet(message)),
  } as Logger;

  (Object.keys(LOG_TYPES) as LogMethods[]).forEach(key => {
    logger[key] = (...args) => log(key, ...args);
  });

  Object.defineProperty(logger, 'level', {
    get: () => maxLevel,
    set(val) {
      maxLevel = val;
    },
  });

  logger.override = customLogger => {
    Object.assign(logger, customLogger);
  };

  return logger;
};
