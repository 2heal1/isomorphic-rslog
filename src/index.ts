import { createLogger } from './createLogger';

export { createLogger } from './createLogger';

export const logger = createLogger();

export type {
  Options,
  LogType,
  LogLevel,
  LogMessage,
  LogFormatter,
  LogFunction,
} from './types';
