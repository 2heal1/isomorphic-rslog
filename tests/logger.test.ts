import { createLogger, logger } from '../src';
import { join } from 'path';

expect.addSnapshotSerializer({
  test: val => Array.isArray(val),
  print: val => {
    const root = join(__dirname, '..');
    return (val as any).toString().replaceAll(root, '<ROOT>');
  },
});

describe('logger', () => {
  test('should log as expected', () => {
    console.log = jest.fn();

    logger.log('this is a log message');
    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.error(new Error('this is an error message'));
    logger.success('this is a success message');

    expect((console.log as jest.Mock).mock.calls).toMatchSnapshot();
  });

  test('should create new logger correctly', () => {
    console.log = jest.fn();

    const logger = createLogger({
      level: 'error',
    });

    logger.log('this is a log message');
    logger.info('this is an info message');
    logger.warn('this is a warn message');
    logger.ready('this is a ready message');
    logger.debug('this is a debug message');
    logger.error(new Error('this is an error message'));
    logger.success('this is a success message');

    expect((console.log as jest.Mock).mock.calls).toMatchSnapshot();
  });
});
