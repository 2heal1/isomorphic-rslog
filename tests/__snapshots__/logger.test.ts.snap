// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger should create new logger with info level correctly 1`] = `
[
  [
    "[1m[31merror  [39m[22m this is a error message",
  ],
  [
    "[1m[36minfo   [39m[22m this is an info message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
  [
    "[1m[32mready  [39m[22m this is a ready message",
  ],
  [
    "[1m[32msuccess[39m[22m this is a success message",
  ],
]
`;

exports[`logger should create new logger with warn level correctly 1`] = `
[
  [
    "[1m[31merror  [39m[22m this is a error message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
]
`;

exports[`logger should log as expected 1`] = `
[
  [
    "[1m[38;2;179;250;236mðŸ˜Š[39m[38;2;179;250;236m [39m[38;2;170;245;228mR[39m[38;2;160;240;221ms[39m[38;2;151;235;213ml[39m[38;2;141;230;206mo[39m[38;2;131;224;199mg[39m[38;2;131;224;199m [39m[38;2;122;219;191mv[39m[38;2;112;214;184m1[39m[38;2;103;209;176m.[39m[38;2;93;204;169m0[39m[38;2;84;199;161m.[39m[38;2;74;194;154m0[39m[38;2;74;194;154m
[39m[22m",
  ],
  [
    "this is a log message",
  ],
  [
    "[1m[36minfo   [39m[22m this is an info message",
  ],
  [
    "[1m[33mwarn   [39m[22m this is a warn message",
  ],
  [
    "[1m[32mready  [39m[22m this is a ready message",
  ],
  [
    "[1m[32msuccess[39m[22m this is a success message",
  ],
]
`;

exports[`logger should log error with stack correctly 1`] = `
[1m[31merror  [39m[22m this is an error message
[90m    at Object.<anonymous> (<ROOT>/tests/logger.test.ts:67:18)
    at Promise.then.completed (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/utils.js:298:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/utils.js:231:10)
    at _callCircusTest (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/run.js:316:40)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at _runTest (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/run.js:252:3)
    at _runTestsForDescribeBlock (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/run.js:126:9)
    at _runTestsForDescribeBlock (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/run.js:121:9)
    at run (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/run.js:71:3)
    at runAndTransformResultsToJestFormat (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapterInit.js:122:21)
    at jestAdapter (<ROOT>/node_modules/.pnpm/jest-circus@29.7.0/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapter.js:79:19)
    at runTestInternal (<ROOT>/node_modules/.pnpm/jest-runner@29.7.0/node_modules/jest-runner/build/runTest.js:367:16)
    at runTest (<ROOT>/node_modules/.pnpm/jest-runner@29.7.0/node_modules/jest-runner/build/runTest.js:444:34)
    at Object.worker (<ROOT>/node_modules/.pnpm/jest-runner@29.7.0/node_modules/jest-runner/build/testWorker.js:106:12)[39m
`;
